src/librustc_mir/lints.rs

```diff
@@ -0,0 +1,130 @@
+
+    while let Some(bb_id) = work_queue.pop() {
+        if visited.contains(bb_id) {
+            //already done
+            continue;
+        }
+
+        visited.insert(bb_id);
+
+        let block = &basic_blocks[bb_id];
+
+        if let Some(ref terminator) = block.terminator {
+            match terminator.kind {
+                TerminatorKind::Call { ref func, .. } => {
+                    let func_ty = func.ty(mir, tcx);
+
+                    if let TyKind::FnDef(fn_def_id, substs) = func_ty.sty {
+                        let (call_fn_id, call_substs) =
+                            if let Some(instance) = Instance::resolve(tcx,
+                                                                        param_env,
+                                                                        fn_def_id,
+                                                                        substs) {
+                                (instance.def_id(), instance.substs)
+                            } else {
+                                (fn_def_id, substs)
+                            };
+
+                        //is this a self call?
+                        let is_self_call =
+                            if let Some(Def::Method(_)) = tcx.describe_def(def_id) {
+                                call_fn_id == def_id && call_substs == caller_substs
+                            } else {
+                                call_fn_id == def_id
+                            };
+
+                        if is_self_call {
+                            self_call_locations.push(terminator.source_info);
+
+                            //this is a self call so we shouldn't explore
+                            //further down this path
+                            continue;
+                        }
+                    }
+                },
+                TerminatorKind::Abort | TerminatorKind::Return => {
+                    //found a path!
+                    reached_exit_without_self_call = true;
+                    break;
+                }
+                _ => {}
+            }
+
+            for successor in terminator.successors() {
+                work_queue.push(*successor);
+            }
+        }
+    }
+
+    // Check the number of self calls because a function that
+    // doesn't return (e.g. calls a `-> !` function or `loop { /*
+    // no break */ }`) shouldn't be linted unless it actually
+    // recurs.
+    if !reached_exit_without_self_call && !self_call_locations.is_empty() {
+        let node_id = tcx.hir.as_local_node_id(def_id).unwrap();
+        let sp = tcx.sess.source_map().def_span(tcx.hir.span(node_id));
+        let mut db = tcx.struct_span_lint_node(UNCONDITIONAL_RECURSION,
+                                                node_id,
+                                                sp,
+                                                "function cannot return without recursing");
```

Comment:
This would be a good opportunity to change this misleading text, along the lines of `function recurses infinitely and so cannot return` (as the function cannot return if it *does* recurse).